components/compositing/compositor.rs

```diff
@@ -1280,107 +1283,122 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
-                },
-                // Leave ScrollLocation unchanged if it is Start or End location.
-                sl @ ScrollLocation::Start | sl @ ScrollLocation::End => sl,
-            };
+        let zoom_changed =
+            self.set_pinch_zoom_level(self.pinch_zoom_level() * combined_magnification);
+        let scroll_result = combined_scroll_event.and_then(|combined_event| {
             let cursor = (combined_event.cursor.to_f32() / self.scale).to_untyped();
-            let cursor = WorldPoint::from_untyped(cursor);
-            let mut txn = Transaction::new();
-
-            let result = match self.hit_test_at_point(cursor) {
-                Some(result) => result,
-                None => return,
-            };
+            self.scroll_node_at_world_point(
+                WorldPoint::from_untyped(cursor),
+                combined_event.scroll_location,
+            )
+        });
+        if !zoom_changed && scroll_result.is_none() {
+            return;
+        }
 
-            if let Some(details) = self.pipeline_details.get_mut(&result.pipeline_id) {
-                match details
-                    .scroll_tree
-                    .scroll_node_or_ancestor(&result.scroll_tree_node, scroll_location)
-                {
-                    Some((external_id, offset)) => {
-                        let scroll_origin = LayoutPoint::new(-offset.x, -offset.y);
-                        txn.scroll_node_with_id(
-                            scroll_origin,
-                            external_id,
-                            ScrollClamping::NoClamping,
-                        );
-                    },
-                    None => {},
-                }
-            }
-            self.send_scroll_positions_to_layout_for_pipeline(&result.pipeline_id);
+        let mut transaction = Transaction::new();
+        if zoom_changed {
+            self.set_root_display_list(&mut transaction);
+        }
 
-            if combined_event.magnification != 1.0 {
-                let old_zoom = self.pinch_zoom_level();
-                self.set_pinch_zoom_level(old_zoom * combined_event.magnification);
-                self.set_root_display_list(&mut txn);
-            }
-            txn.generate_frame(0);
-            self.webrender_api
-                .send_transaction(self.webrender_document, txn);
+        if let Some((pipeline_id, external_id, offset)) = scroll_result {
+            let scroll_origin = LayoutPoint::new(-offset.x, -offset.y);
+            transaction.scroll_node_with_id(scroll_origin, external_id, ScrollClamping::NoClamping);
+            self.send_scroll_positions_to_layout_for_pipeline(&pipeline_id);
             self.waiting_for_results_of_scroll = true
         }
+
+        transaction.generate_frame(0);
+        self.webrender_api
+            .send_transaction(self.webrender_document, transaction);
+    }
+
+    /// Given the result of a hit test and a ScrollLocation try to scroll the node at the result,
```

Comment:
This method doesn't accept the hit test result directly, but accepts the cursor position and performs a hit test. Not sure if this matters, just pointing it out.

Reply:
Thanks for pointing this out. The documentation was totally wrong. I've updated it.